File: SampleApps/ObjC/AdColonyInstantFeed/AdColonyInstantFeed/ViewController.m
//
//  ViewController.m
//  AdColonyInstantFeed
//
//  Copyright (c) 2016 AdColony. All rights reserved.
//

#import "ViewController.h"

#import "FeedCellImage.h"
#import "FeedCellAd.h"

#define kAd             @"ad"
#define kPost           @"post"
#define kAdView         @"adView"
#define kAdZone         @"adZone"
#define kCellType       @"cellType"
#define kPostImage      @"postImage"
#define kPostImageAR    @"aspectRatio"
#define kFeedCellImage  @"FeedCellImage"
#define kFeedCellAd     @"FeedCellAd"
#define kAdColonyAppID  @"app2086517932ad4b608a"
#define kAdColonyZoneID @"vz7c0765ee52af4d67b9"

#define kAdQueueLimit 2

#define kAdViewCellIndex  4
#define kAdViewCellHeight 277.0f

@interface ViewController ()
{
    BOOL _active;
    
    NSMutableArray *_posts;
    NSMutableArray *_ads;
}
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *spinner;
@property (weak, nonatomic) IBOutlet UILabel *loadingLabel;
@end


@implementation ViewController

#pragma mark - UIViewController Overrides

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //Configure AdColony once
    [AdColony configureWithAppID:kAdColonyAppID zoneIDs:@[kAdColonyZoneID] options:nil completion:^(NSArray<AdColonyZone *> * zones) { [self requestAd]; }];
    
    //Hardcoded data source for our feed
    _posts = [@[@{kCellType : kPost, kPostImage : @"Taco-Bell", kPostImageAR : @3.2f},
                @{kCellType : kPost, kPostImage : @"Pacific",   kPostImageAR : @1.78f},
                @{kCellType : kPost, kPostImage : @"MLB",       kPostImageAR : @1.45f},
                @{kCellType : kPost, kPostImage : @"MTV",       kPostImageAR : @1.68f},
                @{kCellType : kPost, kPostImage : @"Fallon",    kPostImageAR : @1.33f},
                @{kCellType : kPost, kPostImage : @"Cashmore",  kPostImageAR : @1.68f},
                @{kCellType : kPost, kPostImage : @"Pugs",      kPostImageAR : @1.41f},
                @{kCellType : kPost, kPostImage : @"Jobs",      kPostImageAR : @1.78f}] mutableCopy];
    
    //Cached ad source
    _ads = [@[] mutableCopy];
    
    //Register our nibs
    [_tableView registerNib:[UINib nibWithNibName:kFeedCellImage bundle:nil] forCellReuseIdentifier:kFeedCellImage];
    [_tableView registerNib:[UINib nibWithNibName:kFeedCellAd bundle:nil] forCellReuseIdentifier:kFeedCellAd];
    
    //Hide the table view until we have at least one ready ad
    _tableView.hidden = YES;
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
        return UIInterfaceOrientationMaskAll;
    } else {
        return UIInterfaceOrientationMaskPortrait | UIInterfaceOrientationMaskPortraitUpsideDown;
    }
}

- (BOOL)shouldAutorotate {
    return YES;
}

#pragma mark - Requesting Ads

- (void)requestAd {
    [AdColony requestNativeAdViewInZone:kAdColonyZoneID size:CGSizeMake(300, 169) options:nil viewController:self
        success:^(AdColonyNativeAdView* _Nonnull ad) {
            NSLog(@"AdColony returned a valid native ad view for zone: %@", ad.zoneID);
            
            //Referencing the ad object from within one of its block-based handlers will cause a retain cycle
            __weak AdColonyNativeAdView* weakAd = ad;
            
            //Native video start handler
            ad.start = ^{
                NSLog(@"AdColonyNativeAdView has started");
            };
            
            //Native video finish handler
            ad.finish = ^{
                NSLog(@"AdColonyNativeAdView finished");
                
                //If we have no more native ads ready, just leave the finished one in the feed
                if ([_ads count] == 0) {
                    return;
                }
                
                //If the native ad has been expanded to fullscreen, use the close handler instead
                if (weakAd.opened) {
                    return;
                }
                
                //Try to get an ad that's ready to be viewed and then try to insert it into our feed
                //*** NOTE: Replacing finished ads with new ones will increase publisher revenue
                [self replaceCurrentAd:weakAd];
            };
            
            //Native video open handler
            ad.open = ^{
                NSLog(@"AdColonyNativeAdView has opened");
            };
            
            //Native ad close handler
            ad.close = ^{
                NSLog(@"AdColonyNativeAdView has closed");
            };
            
            //Try to insert the new ad view into our feed
            //If it is already full, queue the new ad for later use
            if (![self updateFeedWithAdView:ad]) {
                NSLog(@"Feed is full of ads right now. Queuing ad for later use");
                [_ads addObject:ad];
            }
            
            //Unhide the table view if this handler indicates at least one ad is ready
            if (!_active) {
                [_spinner stopAnimating];
                _loadingLabel.hidden = YES;
                _tableView.hidden = NO;
                _active = YES;
            }
            
            //Try to get a new ad
            if ([_ads count] <= kAdQueueLimit) {
                [self requestAd];
            }
        }
        failure:^(AdColonyAdRequestError* _Nonnull error) {
            NSLog(@"AdColony returned an error: %@ with suggestion: %@", [error localizedDescription], [error localizedRecoverySuggestion]);
            [self requestAd];
        }
     ];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [_posts count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSDictionary *cellConfig = _posts[indexPath.row];
    NSString *cellType = cellConfig[kCellType];
    
    //There are only two cases to consider here:
    // 1. The cell cannot be an ad
    // 2. The cell is an ad and we have an AdColonyNativeAdView
    if ([cellType isEqualToString:kPost]) {
        return [self createStandardCell:indexPath]; //The cell is just a regular post, so return a standard FeedCell
    } else {
        AdColonyNativeAdView *adView = (AdColonyNativeAdView *)cellConfig[kAdView]; //The cell is an ad, and we have an ad view already
        return [self createCellWithAdView:adView indexPath:indexPath]; //Return a FeedCellAd object
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSDictionary *cellConfig = _posts[indexPath.row];
    NSString *cellType = cellConfig[kCellType];
    
    //There are only two cases to consider here:
    // 1. The cell cannot be an ad
    // 2. The cell is an ad and we have an AdColonyNativeAdView
    if ([cellType isEqualToString:kPost]) {
        return [self getImageCellHeight:cellConfig]; //The cell is just a regular post, so return the height for a standard FeedCell
    } else {
        return kAdViewCellHeight; //The cell should be an ad view, so return the height for a FeedCellAd object
    }
}

#pragma mark - Feed Cell Creation

- (FeedCellAd *)createCellWithAdView:(AdColonyNativeAdView *)adView indexPath:(NSIndexPath *)indexPath {
    FeedCellAd *adCell = [_tableView dequeueReusableCellWithIdentifier:kFeedCellAd forIndexPath:indexPath];
    
    //Configure the cell's view elements using the properties of the AdColonyNativeAdView
    adCell.advertiserLabel.text = adView.advertiserName;
    adCell.iconView.image  = adView.advertiserIcon;
    adCell.titleLabel.text = adView.adTitle;
    
    //Size the native ad view appropriately
    CGFloat videoWidth  = adCell.adContainer.frame.size.width;
    CGFloat videoHeight = adCell.adContainer.frame.size.height;
    [adView setFrame:CGRectMake(0, 0, videoWidth, videoHeight)];
    
    //Add the video view to the cell
    adCell.adView = adView;
    
    return adCell;
}

- (FeedCellImage *)createStandardCell:(NSIndexPath *)indexPath {
    //Create a standard cell with an image
    NSDictionary *cellConfig = _posts[indexPath.row];
    FeedCellImage *cell = [_tableView dequeueReusableCellWithIdentifier:kFeedCellImage forIndexPath:indexPath];
    cell.background.image  = [UIImage imageNamed:cellConfig[kPostImage]];
    
    return cell;
}

- (CGFloat)getImageCellHeight:(NSDictionary *)cellConfig {
    CGFloat aspectRatio = [cellConfig[kPostImageAR] floatValue];
    CGFloat tableWidth  = _tableView.frame.size.width;
    return tableWidth / aspectRatio;
}

#pragma mark - Updating Feed and Data Source

- (void)replaceCurrentAd:(AdColonyNativeAdView *)currentAd {
    //Try to insert the next ad into our feed
    AdColonyNativeAdView* newAd = _ads[0];
    [self updateFeedWithAdView:newAd];
    
    //Update the queue of waiting ads
    [_ads removeObjectAtIndex:0];
    
    //Destroy the current ad to free up resources
    [currentAd removeFromSuperview];
    [currentAd destroy];
}

- (BOOL)updateFeedWithAdView:(AdColonyNativeAdView *)adView  {
    NSDictionary *cellConfig = _posts[kAdViewCellIndex];
    NSString *cellType = cellConfig[kCellType];
    
    //We want to insert an ad view in the 5th position of the feed if possible
    //If the cell at that position is currently an image, insert an ad view at that index
    if (![cellType isEqualToString:kAd]) {
        [self updateDataSourceWithAdView:adView atIndex:kAdViewCellIndex];
        return YES;
    }
    
    //If the current ad view is finished, replace it with the new one
    //*** NOTE: Replacing finished ads with new ones will increase publisher revenue
    AdColonyNativeAdView *currentAd = cellConfig[kAdView];
    if (currentAd && currentAd.finished) {
        [_posts removeObjectAtIndex:kAdViewCellIndex];
        [self updateDataSourceWithAdView:adView atIndex:kAdViewCellIndex];
        return YES;
    }
    
    return NO;
}

- (void)updateDataSourceWithAdView:(AdColonyNativeAdView *)adView atIndex:(NSUInteger)index {
    //Update our data source with the new ad view and reload the table view
    [_posts insertObject:@{kCellType : kAd, kAdView : adView} atIndex:index];
    [_tableView reloadData];
}

- (void)removeCurrentAdViewFromDataSource {
    //Remove the old ad view from our data source
    NSDictionary *cellConfig = _posts[kAdViewCellIndex];
    AdColonyNativeAdView *currentAd = cellConfig[kAdView];
    if (currentAd) {
        [_posts removeObjectAtIndex:kAdViewCellIndex];
    }
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    //Resume the ad if our table view is going to display it
    if ([cell isKindOfClass:[FeedCellAd class]]) {
        FeedCellAd* adCell = (FeedCellAd*)cell;
        NSLog(@"Resuming ad view");
        [adCell resume];
    }
}

- (void)tableView:(UITableView *)tableView didEndDisplayingCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    //Pause the ad if our table view has finished displaying it
    if ([cell isKindOfClass:[FeedCellAd class]]) {
        FeedCellAd* adCell = (FeedCellAd*)cell;
        NSLog(@"Pausing ad view");
        [adCell pause];
    }
}
@end

File: SampleApps/Swift/AdColonyInstantFeed/AdColonyInstantFeed/ViewController.swift
//
//  ViewController.swift
//  AdColonyInstantFeed
//
//  Copyright (c) 2016 AdColony. All rights reserved.
//

import UIKit

struct Constants
{
    static let ad = "ad"
    static let post = "post"
    static let adView = "adView"
    static let adZone = "zone"
    static let cellType = "cellType"
    static let postImage = "postImage"
    static let postImageAR = "aspectRatio"
    
    static let feedCellImage = "FeedCellImage"
    static let feedCellAd = "FeedCellAd"
    
    static let adColonyAppID = "app2086517932ad4b608a"
    static let adColonyZoneID = "vz7c0765ee52af4d67b9"
    
    static let adViewCellIndex = 4
    static let adQueueLimit = 2
    
    static let adViewWidth:CGFloat = 300.0
    static let adViewHeight:CGFloat = 169.0
    static let adViewCellHeight:CGFloat = 277.0
}


class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate
{
    @IBOutlet weak var spinner: UIActivityIndicatorView!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var loadingLabel: UILabel!
    
    var posts = [[String : Any?]]()
    var ads = [AdColonyNativeAdView]()
    
    var active: Bool = false
    
    
    //===================================================
    // MARK:- UIViewController Overrides
    //===================================================
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        //Configure AdColony once
        AdColony.configure(withAppID: Constants.adColonyAppID, zoneIDs: [Constants.adColonyZoneID], options: nil, completion: { (zones) in self.requestAd() })
        
        //Hardcoded data source for our feed
        posts = [[Constants.cellType : Constants.post, Constants.postImage : "Taco-Bell", Constants.postImageAR : 3.2 as CGFloat],
                 [Constants.cellType : Constants.post, Constants.postImage : "Pacific",   Constants.postImageAR : 1.78 as CGFloat],
                 [Constants.cellType : Constants.post, Constants.postImage : "MLB",       Constants.postImageAR : 1.45 as CGFloat],
                 [Constants.cellType : Constants.post, Constants.postImage : "MTV",       Constants.postImageAR : 1.68 as CGFloat],
                 [Constants.cellType : Constants.post, Constants.postImage : "Fallon",    Constants.postImageAR : 1.33 as CGFloat],
                 [Constants.cellType : Constants.post, Constants.postImage : "Jobs",      Constants.postImageAR : 1.7 as CGFloat],
                 [Constants.cellType : Constants.post, Constants.postImage : "Pugs",      Constants.postImageAR : 1.41 as CGFloat],
                 [Constants.cellType : Constants.post, Constants.postImage : "Cashmore",  Constants.postImageAR : 1.68 as CGFloat]]
        
        //Register our nibs
        tableView.register(UINib(nibName: Constants.feedCellImage, bundle: nil), forCellReuseIdentifier: Constants.feedCellImage)
        tableView.register(UINib(nibName: Constants.feedCellAd, bundle: nil), forCellReuseIdentifier: Constants.feedCellAd)
        
        //Hide the table view until we have at least one ready ad
        tableView.isHidden = true
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle
    {
        return UIStatusBarStyle.lightContent
    }
    
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask
    {
        if UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiom.pad {
            return UIInterfaceOrientationMask.all
        } else {
            return UIInterfaceOrientationMask.portrait
        }
    }
    
    override var shouldAutorotate: Bool
    {
        return true
    }
    
    
    //===================================================
    // MARK:- Requesting Ads
    //===================================================
    
    func requestAd()
    {
        AdColony.requestNativeAdView(inZone: Constants.adColonyZoneID, size: CGSize(width: 300, height: 170), options: nil, viewController: self,
            success:{ (adView) in
                NSLog("AdColony returned a valid native ad view for zone: " + adView.zoneID)
                
                //Native video start handler
                adView.setStart({ _ in
                    NSLog("AdColonyNativeAdView video started")
                })
                
                //Native video finish handler
                adView.setFinish({ _ in
                    NSLog("AdColonyNativeAdView video finished")
                        
                    //If the native ad has been expanded to fullscreen, use the close handler instead
                    guard !adView.opened else { return }
                    
                    self.onAdComplete(adView)
                })
                
                //Native video open handler
                adView.setOpen({ _ in
                    NSLog("AdColonyNativeAdView opened")
                })
                
                //Native ad close handler
                adView.setClose({ _ in
                    NSLog("AdColonyNativeAdView closed")
                    self.onAdComplete(adView)
                })
                
                //Try to insert the new ad view into our feed
                //If it is already full, queue the new ad for later use
                if !self.updateFeedWithAdView(adView)
                {
                    NSLog("Feed is full of ads right now. Queuing ad for later use");
                    self.ads.append(adView)
                }
                
                //Unhide the table view if this handler indicates at least one ad is ready
                if !self.active {
                    self.spinner.stopAnimating()
                    self.loadingLabel.isHidden = true
                    self.tableView.isHidden = false
                    self.active = true
                }
                
                //Try to get a new ad
                if self.ads.count <= Constants.adQueueLimit {
                    self.requestAd()
                }
            },
            failure:{ (error) in
                NSLog("AdColony returned an error: " + error.localizedDescription + " with suggestion: " + error.localizedRecoverySuggestion!)
                self.requestAd();
            }
        )
    }
    
    func onAdComplete(_ ad: AdColonyNativeAdView)
    {
        //If we have no more ads ready, just leave the finished one in the feed
        guard self.ads.count != 0 else { return }
        
        //Try to get an ad that's ready to be viewed and then try to insert it into our feed
        //*** NOTE: Replacing finished ads with new ones will increase publisher revenue
        guard let newAd = self.ads.first else { return }
        
        //Try to insert the next ad into our feed
        if self.updateFeedWithAdView(newAd) {
            //Update the queue of waiting ads
            self.ads.remove(at: 0)
            
            //Destroy the current ad to free up resources
            ad.removeFromSuperview()
            ad.destroy()
        }
    }
    
    //===================================================
    // MARK:- UITableViewDataSource
    //===================================================
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return posts.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        let cellConfig = posts[(indexPath as NSIndexPath).row]
        let cellType = cellConfig[Constants.cellType] as! String
        
        //There are only two cases to consider here:
        // 1. The cell cannot be an ad
        // 2. The cell is an ad and we have an AdColonyNativeAdView
        if cellType == Constants.post {
            return self.createStandardCell(indexPath)
        } else {
            let adView = cellConfig[Constants.adView] as! AdColonyNativeAdView
            return self.createCellWithAdView(adView, indexPath: indexPath)
        }
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat
    {
        let cellConfig = posts[(indexPath as NSIndexPath).row]
        let cellType = cellConfig[Constants.cellType] as! String
        
        //There are only two cases to consider here:
        // 1. The cell cannot be an ad
        // 2. The cell is an ad and we have an AdColonyNativeAdView
        if cellType == Constants.post {
            return self.getStandardCellHeight(cellConfig)
        } else {
            return Constants.adViewCellHeight
        }
    }
    
    
    //===================================================
    // MARK:- Feed Cell Creation
    //===================================================
    
    func createCellWithAdView(_ adView: AdColonyNativeAdView, indexPath: IndexPath) -> FeedCellAd
    {
        let adCell = tableView.dequeueReusableCell(withIdentifier: Constants.feedCellAd, for: indexPath) as! FeedCellAd
        
        //Configure the cell's view elements using the properties of the AdColonyNativeAdView
        adCell.advertiserLabel.text = adView.advertiserName
        adCell.iconView.image  = adView.advertiserIcon
        adCell.titleLabel.text = adView.adTitle

        //Size the native ad view appropriately
        adView.frame = CGRect(x: 0, y: 0, width: Constants.adViewWidth, height: Constants.adViewHeight)
        
        //Configure the native ad's engagement button
        adView.engagementButton?.backgroundColor = UIColor.black
        
        //Add the video view to the cell
        adCell.adView = adView
        
        return adCell
    }
    
    func createStandardCell(_ indexPath: IndexPath) -> FeedCellImage
    {
        //Create a standard cell with an image
        let cellConfig = posts[(indexPath as NSIndexPath).row]
        let cell = tableView.dequeueReusableCell(withIdentifier: Constants.feedCellImage, for: indexPath) as! FeedCellImage
        if let image = UIImage(named: cellConfig[Constants.postImage] as! String) {
            cell.background.image = image
        }
        
        return cell
    }
    
    func getStandardCellHeight(_ cellConfig: Dictionary<String, Any?>) -> CGFloat
    {
        let aspectRatio = cellConfig[Constants.postImageAR] as! CGFloat
        let tableWidth  = tableView.frame.size.width
        return tableWidth / aspectRatio
    }
    
    
    //===================================================
    // MARK:- Updating Feed and Data Source
    //===================================================
    
    func updateFeedWithAdView(_ adView: AdColonyNativeAdView) -> Bool
    {
        let cellConfig = posts[Constants.adViewCellIndex]
        let cellType = cellConfig[Constants.cellType] as! String
        
        //We want to insert an ad view in the 5th position of the feed if possible
        //If the cell at that position is currently an image, insert an ad view at that index
        if cellType != Constants.ad {
            self.updateDataSourceWithAdView(adView, index: Constants.adViewCellIndex)
            return true
        }
        
        //If the current ad view is finished, replace it with the new one
        //*** NOTE: Replacing finished ads with new ones will increase publisher revenue
        if let oldAdView = cellConfig[Constants.adView] as? AdColonyNativeAdView {
            if oldAdView.finished {
                posts.remove(at: Constants.adViewCellIndex)
                self.updateDataSourceWithAdView(adView, index: Constants.adViewCellIndex)
                return true
            }
        }
        
        return false
    }
    
    func updateDataSourceWithAdView(_ adView: AdColonyNativeAdView, index: Int)
    {
        //Update our data source with the new ad view and reload the table view
        posts.insert([Constants.cellType : Constants.ad, Constants.adView : adView], at: index)
        tableView.reloadData()
    }
    
    func removeCurrentAdViewFromDataSource()
    {
        //Remove the old ad view from our data source
        let cellConfig = posts[Constants.adViewCellIndex]
        if let _ = cellConfig[Constants.adView] {
            posts.remove(at: Constants.adViewCellIndex)
        }
    }
    
    
    //===================================================
    // MARK:- UITableViewDelegate
    //===================================================
    
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath)
    {
        guard let adCell = cell as? FeedCellAd else { return }
        
        //Resume the ad if our table view is going to display it
        adCell.adView?.resume()
    }
    
    func tableView(_ tableView: UITableView, didEndDisplaying cell: UITableViewCell, forRowAt indexPath: IndexPath)
    {
        guard let adCell = cell as? FeedCellAd else { return }
        
        //Pause the ad if our table view is going to display it
        adCell.adView?.pause()
    }
}
